### ISSUE TITLE
[jit] Can't index nn.ModuleList in script function

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#16123]
## 🐛 Bug

Can't index nn.ModuleList in script function

## To Reproduce

```

import torch

class Model(torch.jit.ScriptModule):
    def __init__(self, ):
        super(Model, self).__init__()
        self.layers = torch.nn.ModuleList([torch.nn.Linear(2, 2)])

    @torch.jit.script_method
    def forward(self, input, index):
        return self.layers[index](input)

model = Model()
model(torch.randn(2, 2))
```

provides the following trace:
```
RuntimeError:
python value of type 'ModuleList' cannot be used as a value:
@torch.jit.script_method
def forward(self, input, index):
    return self.layers[index](input)
           ~~~~~~~~~~~ <--- HERE
```

## Expected behavior

No error

## Environment
Pytorch 1.0.0

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @gmagogsfm @suo

**Comment by bswrundquist** [comment]
Even if you define \_\_contants__ to use the nn.ModuleList like in example in documents it still errors out (little more than halfway down).
https://pytorch.org/docs/stable/jit.html?highlight=model%20features

Here is code (changed super for first class call from example so it works)

```python
import torch
from torch import nn

class SubModule(torch.jit.ScriptModule):
    def __init__(self):
        super(SubModule, self).__init__()
        self.weight = nn.Parameter(torch.randn(2))

    @torch.jit.script_method
    def forward(self, input):
        return self.weight + input

class MyModule(torch.jit.ScriptModule):
    __constants__ = ['mods']

    def __init__(self):
        super(MyModule, self).__init__()
        self.mods = torch.nn.ModuleList([SubModule() for i in range(10)])

    @torch.jit.script_method
    def forward(self, v):
        for module in self.mods:
            v = m(v)
        return v
MyModule()
```
produces 
```
RuntimeError: 
undefined value m:
@torch.jit.script_method
def forward(self, v):
    for module in self.mods:
        v = m(v)
            ~ <--- HERE
    return v
```
## Environment:
PyTorch 1.0.0

**Comment by zou3519** [comment]
@bswrundquist it should work, you have a typo:

```
    @torch.jit.script_method
    def forward(self, v):
        for module in self.mods:
            v = m(v)
```

should be 

```
    @torch.jit.script_method
    def forward(self, v):
        for module in self.mods:
            v = module(v)
```

**Comment by bswrundquist** [comment]
Good catch, I didn't see that when I copied from documentation.

**Comment by bswrundquist** [comment]
@zou3519 Thanks for the correction.

**Comment by Evpok** [comment]
Iterating over the ModuleList does work, but directly indexing it still doesn't, which is to be expected, since ModuleList can't be indexed in non-jit code either.

**Comment by sidazhang** [comment]
@zou3519 Do you know if the PyTorch team has a timeline for fixing this? I have a module that has a lot of indexing into ModuleList. I am trying to figure out if I should wait for the fix. (Because, it is kind of tedious to rewrite all these code - so I am hoping to defer it for as long as possible)

**Comment by zou3519** [comment]
I remember the argument against having this is that if we don't statically know the size of the ModuleList then indexing into it isn't possible. However, this might be possible if we mark the ModuleList as a part of the `__constants__`... (@eellison @suo can one of you chime in here?)

**Comment by Evpok** [comment]
@zou3519 If I understand correctly, you can't refer to the `ModuleList` if it is not in the constants anyway.

**Comment by apaszke** [comment]
@sidazhang I agree `enumerate` is pretty important and we should support it, but if you need a quick workaround here's poor man's enumerate you could use:
```
i = 0
for mod in self.modules:
   mod(output[i])
   i += 1
```

**Comment by apaszke** [comment]
@zou3519 I think we won't be able to implement indexing because modules/functions are not first class values in TorchScript, so you can't assign them to variables (ok, you can, but we need to be able to resolve them statically). So this is a much more fundamental difficulty that would require implementing a proper call mechanism. The requirement is there today because the only form of calling we support is function inlining.

**Comment by sidazhang** [comment]
> @sidazhang I agree `enumerate` is pretty important and we should support it, but if you need a quick workaround here's poor man's enumerate you could use:
> 
> ```python
> i = 0
> for mod in self.modules:
>    mod(output[i])
>    i += 1
> ```

Ehh. This is kind of gross. lol

Ok. I can do that

**Comment by DanlanChen** [comment]
Please fix it! I have this problem as well, it is not working for Music Vae

**Comment by DanlanChen** [comment]
> @sidazhang I agree `enumerate` is pretty important and we should support it, but if you need a quick workaround here's poor man's enumerate you could use:
> 
> ```python
> i = 0
> for mod in self.modules:
>    mod(output[i])
>    i += 1
> ```

It is not working

**Comment by suo** [comment]
@Danlanchen can you give us more information about why it’s not working? Like an error msg and a reproducible way to get the error. Thanks!

**Comment by DanlanChen** [comment]
Sorry I fix it my adding self.modules in __constants__ @suo

**Comment by DanlanChen** [comment]
Could we use `zip` if we want to iterate through two modulelist?
> ```python
i = 0
for mod1, mod2 in zip(self.modules1, self.module2):
   x = mod1(output[i])
   x = mod2(x)
   i += 1

**Comment by eellison** [comment]
Is this possible yet with first class modules?

**Comment by TheCodez** [comment]
The following piece of code is still not working:

```python
class MyMod(torch.jit.ScriptModule):
    __constants__ = ['loc_layers']
    def __init__(self):
        super(MyMod, self).__init__()
        locs = []
        for i in range(4):
            locs.append(nn.Conv2d(64, 21, kernel_size=1))
        self.loc_layers = nn.ModuleList(locs)

    @torch.jit.script_method
    def forward(self, input):
        # type: (List[Tensor]) -> Tensor
        locs = []
        i = 0
        for layer in input:
            loc = self.loc_layers[i](layer)
            locs.append(loc)
            i += 1
        loc_preds = torch.cat(locs, 1)
        return loc_preds
```

Gives the following error:
```
module cannot be used as a value:
...
loc = self.loc_layers[i](layer)
      ~~~~~~~~ <--- HERE
```

**Comment by LCWdmlearning** [comment]
> The following piece of code is still not working using 1.1.0:
> 
> ```python
> class MyMod(torch.jit.ScriptModule):
>     __constants__ = ['loc_layers']
>     def __init__(self):
>         super(MyMod, self).__init__()
>         locs = []
>         for i in range(4):
>             locs.append(nn.Conv2d(64, 21, kernel_size=1))
>         self.loc_layers = nn.ModuleList(locs)
> 
>     @torch.jit.script_method
>     def forward(self, input):
>         # type: (List[Tensor]) -> Tensor
>         locs = []
>         i = 0
>         for layer in input:
>             loc = self.loc_layers[i](layer)
>             locs.append(loc)
>             i += 1
>         loc_preds = torch.cat(locs, 1)
>         return loc_preds
> ```
> 
> Gives the following error:
> 
> ```
> module cannot be used as a value:
> ...
> loc = self.loc_layers[i](layer)
>       ~~~~~~~~ <--- HERE
> ```

hello ! I meet same problems.Have you solved it?

**Comment by blythed** [comment]
Any progress here? I'm getting the same errors using a module list.

**Comment by driazati** [comment]
This issue is blocked by some other in-progress work to make `Module` a first-class object in the TorchScript interpreter, until that's finished @TheCodez's solution above is the recommended workaround

**Comment by houseroad** [comment]
Another data point, i also hit this issue... and I have two lists... using `for xxx in module_list` will be painful...

**Comment by sidazhang** [comment]
@driazati being able to zip over 2 ModueLists that would be awesome too

**Comment by ShangxuanWu** [comment]
Another data point, i also hit this issue... and I have two lists... using for xxx in module_list will be painful...

**Comment by marchss** [comment]
Same problem here.

**Comment by suo** [comment]
Just wanted to give a quick update here: this is a high priority item for us, but the fix is not easy given the way TorchScript is implemented so it will take some time. Thanks for the reports, everyone!

**Comment by standbyme** [comment]
#29236 May be helpful.

**Comment by mohammedayub44** [comment]
@suo  Not sure if this has been fixed. I'm running into the same error on a VGG16 model. I'm using the latest Pytorch `master` branch. 

Code snippet:

```python3
class VGG16_frontend(nn.Module):
    def __init__(self,block_num=5,decode_num=0,load_weights=True,bn=False,IF_freeze_bn=False):
        super(VGG16_frontend,self).__init__()
        self.block_num = block_num
        self.load_weights = load_weights
        self.bn = bn
        self.IF_freeze_bn = IF_freeze_bn
        self.decode_num = decode_num

        block_dict = [[64, 64, 'M'], [128, 128, 'M'], [256, 256, 256, 'M'],\
             [512, 512, 512,'M'], [512, 512, 512,'M']]

        self.frontend_feat = []
        for i in range(block_num):
            self.frontend_feat += block_dict[i]

        if self.bn:
            self.features = make_layers(self.frontend_feat, batch_norm=True)
        else:
            self.features = make_layers(self.frontend_feat, batch_norm=False)


        if self.load_weights:
            if self.bn:
                pretrained_model = models.vgg16_bn(pretrained = True)
            else:
                pretrained_model = models.vgg16(pretrained = True)
            pretrained_dict = pretrained_model.state_dict()
            model_dict = self.state_dict()
            # filter out unnecessary keys
            pretrained_dict = {k: v for k, v in pretrained_dict.items() if k in model_dict}
            # overwrite entries in the existing state dict
            model_dict.update(pretrained_dict) 
            # load the new state dict
            self.load_state_dict(model_dict)

        if IF_freeze_bn:
            self.freeze_bn()
    
    def forward(self,x):
        if self.bn: 
            x = self.features[ 0:7](x)
            print(type(x))
            print(x.shape)
            conv1_feat =x if self.decode_num>=4 else []
            x = self.features[ 7:14](x)
            conv2_feat =x if self.decode_num>=3 else []
            x = self.features[ 14:24](x)
            conv3_feat =x if self.decode_num>=2 else []
            x = self.features[ 24:34](x)
            conv4_feat =x if self.decode_num>=1 else []
            x = self.features[ 34:44](x)
            conv5_feat =x 
        else:
            x = self.features[ 0: 5](x)
            conv1_feat =x if self.decode_num>=4 else []
            x = self.features[ 5:10](x)
            conv2_feat =x if self.decode_num>=3 else []
            x = self.features[ 10:17](x)
            conv3_feat =x if self.decode_num>=2 else []
            x = self.features[ 17:24](x)
            conv4_feat =x if self.decode_num>=1 else []
            x = self.features[ 24:31](x)
            conv5_feat =x 
               
        feature_map = {'conv1':conv1_feat,'conv2': conv2_feat,\
            'conv3':conv3_feat,'conv4': conv4_feat, 'conv5': conv5_feat}   
        
        # feature_map = [conv1_feat, conv2_feat, conv3_feat, conv4_feat, conv5_feat]
        
        return feature_map


    def freeze_bn(self):
        for m in self.modules():
            if isinstance(m, nn.BatchNorm2d):
                m.eval()
```

Error while running:

```python

RuntimeError: 
Arguments for call are not valid.
The following variants are available:
  
  aten::slice.Tensor(Tensor(a) self, int dim=0, int start=0, int end=9223372036854775807, int step=1) -> (Tensor(a)):
  Expected a value of type 'Tensor' for argument 'self' but instead found type '__torch__.torch.nn.modules.container.Sequential'.
  
  aten::slice.str(str string, int start, int end=9223372036854775807, int step=1) -> (str):
  Expected a value of type 'str' for argument 'string' but instead found type '__torch__.torch.nn.modules.container.Sequential'.
  
  aten::slice.t(t[] l, int start, int end=9223372036854775807, int step=1) -> (t[]):
  Could not match type __torch__.torch.nn.modules.container.Sequential to List[t] in argument 'l': Cannot match List[t] to __torch__.torch.nn.modules.container.Sequential.

The original call is:
  File "/home/ubuntu/mayub/Github/SS-DCNet/Network/SSDCNet.py", line 92
    def forward(self,x):
        if self.bn: 
            x = self.features[0:7](x)
                ~~~~~~~~~~~~~~~~~ <--- HERE
            conv1_feat =x if self.decode_num>=4 else []
            x = self.features[ 7:14](x)


```


Thanks !

**Comment by suo** [comment]
The current state of things is that we support indexing into a ModuleList with an integer literal, much like we do with tuples. We don't support slicing, which involves creating a new ModuleList. This is theoretically possible; mind filing a separate issue with your example?

**Comment by mohammedayub44** [comment]
Thanks. I have made a new one #38034

**Comment by snie2012** [comment]
Any updates on supporting ModuleDict indexing?

**Comment by kwanUm** [comment]
> @LCWdmlearning using my code as an example you can iterate over the module list instead of the tensor list:
> 
> ```python
> i = 0
> for layer in self.loc_layers:
>     loc = layer(input[i])
>     locs.append(loc)
>     i += 1
> ```

Did that work for anyone? I still get the "'__torch__.torch.nn.modules.container.ModuleList' object is not iterable" when using this (latest pytorch version).

**Comment by Luonic** [comment]
@kwanUm yes I JIT'ed HRNetV2 with this method

**Comment by gemfield** [comment]
on PyTorch 1.6.0, do index on ModuleList still get this error.

**Comment by kwanUm** [comment]
Works for me actually on pytorch 1.6. Make sure you are not iterating on a
ModuleList that is a submodule/data member of the class you work on.
There's an open bug on that.

**Comment by ShangxuanWu** [comment]
Under PyTorch 1.6, `ModuleDict` string indexing is still giving me `Only ModuleList, Sequential, and ModuleDict modules are subscriptable:` error. Is there any ETA for fixing this? If we cannot using string to index dict, this basically defeats the purpose of dict right?

**Comment by CODEJIN** [comment]
Is this problem solved at 1.7.0? I faced same problem by using ModuleDict. :(

**Comment by yangsenius** [comment]
> @kwanUm yes I JIT'ed HRNetV2 with this method

Could you share the JITTed HRNet version?

**Comment by Thien223** [comment]
I have to convert a pretrained pytorch model to TorchScript for deployment on AWS Elastic Inference.

The original model has a loop which zipped 3 modulelists. Stocked here for 2 months.

I have tried to split into 3 nested loop, but it will results Segmentation Fault when I doing like this:
```
        for in_layer in self.modulelist_1:
            idx_1=0
            print(in_layer)
            for cond_layer in self.modulelist_2:
                if idx_2!=idx_1:
                    idx_2+=1 #for making sure 2nd modulelist index is same as 1st modulelist index
                    continue
                else:
                    idx_3=0
                    for res_skip in self.modulelist_3:
                        if idx_3!=idx_1:
                            idx_3+=1 #for making sure 3rd modulelist index is same as 1st and 2nd modulelist index
                            continue
                        else:
                            # when all 3 indexes are same (ex: [1,1,1], [2,2,2]...)
                            #run code here#
```

Now i have to find out the way to loop over 3 modulelists that are not supported zipping, indexing and nested looping.

Anyone has any idea please help. Thank you so much.
